<h1>Tabla de contenidos<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introducción" data-toc-modified-id="Introducción-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introducción</a></span></li><li><span><a href="#Manejo-de-archivos" data-toc-modified-id="Manejo-de-archivos-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Manejo de archivos</a></span><ul class="toc-item"><li><span><a href="#Archivos-y-ruta-de-archivos" data-toc-modified-id="Archivos-y-ruta-de-archivos-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Archivos y ruta de archivos</a></span></li><li><span><a href="#Backslash-en-Windows-y-slash-en-Linux" data-toc-modified-id="Backslash-en-Windows-y-slash-en-Linux-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span><em>Backslash</em> en Windows y <em>slash</em> en Linux</a></span></li><li><span><a href="#El-directorio-de-trabajo-actual" data-toc-modified-id="El-directorio-de-trabajo-actual-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>El directorio de trabajo actual</a></span></li><li><span><a href="#Rutas-absolutas-vs-rutas-relativas" data-toc-modified-id="Rutas-absolutas-vs-rutas-relativas-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Rutas absolutas vs rutas relativas</a></span></li><li><span><a href="#Creando-directorios-nuevos-con-os.makedirs()" data-toc-modified-id="Creando-directorios-nuevos-con-os.makedirs()-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Creando directorios nuevos con <code>os.makedirs()</code></a></span></li></ul></li><li><span><a href="#El-módulo-os.path()" data-toc-modified-id="El-módulo-os.path()-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>El módulo <code>os.path()</code></a></span><ul class="toc-item"><li><span><a href="#Manejando-rutas-absolutas-y-rutas-relativas" data-toc-modified-id="Manejando-rutas-absolutas-y-rutas-relativas-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Manejando rutas absolutas y rutas relativas</a></span></li><li><span><a href="#Encontrando-tamaños-de-archivo-y-el-contenido-de-los-directorios" data-toc-modified-id="Encontrando-tamaños-de-archivo-y-el-contenido-de-los-directorios-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Encontrando tamaños de archivo y el contenido de los directorios</a></span></li><li><span><a href="#Verificando-la-validez-de-la-ruta" data-toc-modified-id="Verificando-la-validez-de-la-ruta-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Verificando la validez de la ruta</a></span></li></ul></li><li><span><a href="#El-proceso-de-lectura/escritura-de-archivos" data-toc-modified-id="El-proceso-de-lectura/escritura-de-archivos-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>El proceso de lectura/escritura de archivos</a></span><ul class="toc-item"><li><span><a href="#Abriendo-archivos-con-la-función-open()" data-toc-modified-id="Abriendo-archivos-con-la-función-open()-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Abriendo archivos con la función <code>open()</code></a></span></li><li><span><a href="#Leyendo-el-contenido-de-los-archivos" data-toc-modified-id="Leyendo-el-contenido-de-los-archivos-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Leyendo el contenido de los archivos</a></span></li><li><span><a href="#Escribiendo-archivos" data-toc-modified-id="Escribiendo-archivos-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Escribiendo archivos</a></span></li></ul></li><li><span><a href="#Guardando-variables-con-el-método-shelve" data-toc-modified-id="Guardando-variables-con-el-método-shelve-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Guardando variables con el método <code>shelve</code></a></span></li><li><span><a href="#Guardando-variables-con-la-función-pprint.pformat()" data-toc-modified-id="Guardando-variables-con-la-función-pprint.pformat()-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Guardando variables con la función <code>pprint.pformat()</code></a></span></li><li><span><a href="#Proyecto:-Generando-archivos-aleatorios-para-evaluaciones" data-toc-modified-id="Proyecto:-Generando-archivos-aleatorios-para-evaluaciones-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Proyecto: Generando archivos aleatorios para evaluaciones</a></span><ul class="toc-item"><li><span><a href="#Paso1:-Almacenar-los-datos-del-quiz-en-un-diccionario" data-toc-modified-id="Paso1:-Almacenar-los-datos-del-quiz-en-un-diccionario-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Paso1: Almacenar los datos del quiz en un diccionario</a></span></li><li><span><a href="#Paso-2:-Crear-el-archivo-del-quiz-y-ordenar-aleatoriamente-las-preguntas" data-toc-modified-id="Paso-2:-Crear-el-archivo-del-quiz-y-ordenar-aleatoriamente-las-preguntas-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Paso 2: Crear el archivo del quiz y ordenar aleatoriamente las preguntas</a></span></li><li><span><a href="#Paso-3:-Crear-las-opciones-de-respuesta" data-toc-modified-id="Paso-3:-Crear-las-opciones-de-respuesta-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Paso 3: Crear las opciones de respuesta</a></span></li><li><span><a href="#Paso-4:-Escribir-el-contenido-de-los-archivos-del-quiz-y-el-archivo-de-respuestas" data-toc-modified-id="Paso-4:-Escribir-el-contenido-de-los-archivos-del-quiz-y-el-archivo-de-respuestas-7.4"><span class="toc-item-num">7.4&nbsp;&nbsp;</span>Paso 4: Escribir el contenido de los archivos del quiz y el archivo de respuestas</a></span></li></ul></li><li><span><a href="#Proyecto:-Multiclipboard" data-toc-modified-id="Proyecto:-Multiclipboard-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Proyecto: Multiclipboard</a></span><ul class="toc-item"><li><span><a href="#Paso-1:-Comentarios-y-configuración-del-shelf" data-toc-modified-id="Paso-1:-Comentarios-y-configuración-del-shelf-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>Paso 1: Comentarios y configuración del shelf</a></span></li><li><span><a href="#Paso-2:-Guardar-el-contenido-del-portapapeles-con-una-palabra-clave" data-toc-modified-id="Paso-2:-Guardar-el-contenido-del-portapapeles-con-una-palabra-clave-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>Paso 2: Guardar el contenido del portapapeles con una palabra clave</a></span></li><li><span><a href="#Paso-3:-Listar-las-palabras-clave-y-cargar-el-contenido-de-la-palabra-clave" data-toc-modified-id="Paso-3:-Listar-las-palabras-clave-y-cargar-el-contenido-de-la-palabra-clave-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>Paso 3: Listar las palabras clave y cargar el contenido de la palabra clave</a></span></li></ul></li></ul></div>

# Introducción

Las variables son una buena forma de almacenar datos mientras nuestros programas están corriendo, pero si deseamos que nuestros datos perduren incluso después de que termine la ejecución del programa, necesitamos guardarlos  en un archivo. Podemos pensar en el contenido de un archivo como un único valor tipo cadena de caracteres potencialmente con tamaño en gigabytes. En este capítulo aprenderemos a usar python para crear, leer y guardar archivos en el disco duro.

# Manejo de archivos

## Archivos y ruta de archivos

Un archivo tiene dos propiedades clave: un *filename* (usualmente escrito en un palabra) y un *path*. El path o ruta, especifica la ubicación de un archivo en el computador. Por ejemplo, en mi portatil tengo un archivo con filename `project.docx` con path `/home/thomy/Documentos`. En Windows ese mismo archivo luciría así: `C:\Users\thomy\Documentos\Project.docx`.

La parte del filename después del punto se conoce como extensión y nos dice el tipo de archivo. `project.docx` es un documento de MS Word. Por otro lado, `home`, `thomy` y `Documentos` hacen referencia a *folders* o directorios. Los directorios pueden contener archivos y/o otros directorios. Por ejemplo, `projects.docx` se encuentra dentro del directorio `Documentos`, éste a su vez se encuentra dentro del directorio `thomy`, quien a su vez se encuentra en el directorio `home`. 

La parte `C:\` del path se le conoce como el *directorio raíz* o *root folder*. En windows, el directorio raíz es también llamado *C:\ drive*. En Linux el directorio raíz es `/`. Otros puertos como el USB u otras unidades como el DVD, aparecen en Windows con nuevas letras de directorios raíz como `D:\` o `E:\`. En Linux aparecen en nuevas directorios como `/mnt` (*mount*). Así mismo, en Sistemas Operativos (SO) Unix/Linux, los nombres son sensibles a mayúsculas y minúsculas.

## *Backslash* en Windows y *slash* en Linux

En windows, las rutas son escritas usando backslash (`\`) para separar los nombres de los directorios. En SO Unix/Linux,  sin embargo, se usa el slash (`/`). Si deseamos que nuestros programas funcionen en cualquier SO, debemos escribir nuestros scripts en Python con la capacidad de manejar ambos casos.
Afortunadamente, esto es fácil de implementar con la función `os.path.join()`. Si pasamos el valor string de los nombres de los diferentes directorios y los archivos individuales en nuestra ruta, `os.path.join()` retornará una cadena con el separador adecuado. Veamos el siguiente ejemplo:

In [2]:
import os
os.path.join('home', 'thomy', 'Documents')

'home/thomy/Documents'

En Windows `os.path.join()` retornará `usr\\bin\\spam`. Nótese que los backslashes son dobles porque cada uno necesita el caracter de escape.

La función `os.path.join()` es útil cuando deseamos crear cadenas para nombres de archivos. Estas cadenas serán pasadas a varias de las funciones que veremos en este módulo. Por ejemplo, el siguiente ejemplo une los nombres de una lista de filenames al final del nombre del directorio.

In [3]:
myFiles = ['accounts.txt', 'details.csv', 'invite.docx']

for filename in myFiles:
    print(os.path.join('C:\\Users\\thomy\\Docuementos', filename))

C:\Users\thomy\Docuementos/accounts.txt
C:\Users\thomy\Docuementos/details.csv
C:\Users\thomy\Docuementos/invite.docx


## El directorio de trabajo actual

Cada programa que corre en nuestro computador tiene un *directorio de trabajo actual* o *cwd* (en inglés Current Working Directory). Cualquier nombre de archivo o ruta que no empiece con el directorio raíz, se asume que se encuentra el directorio de trabajo actual. Podemos obtener el directorio de trabajo actual como cadena de caracteres mediante la función `os.getcwd()` y la podemos cambiar con la función `os.chdir()`. Veamos el siguiente ejemplo.

In [8]:
import os
os.getcwd()

'/home/thomy/Google Drive/ITM/Materias/CursoPythonEgresados/JupiterNotebooks/Proyectos/JupyterNotebooks'

In [10]:
os.chdir('/home/thomy/Documents/')
os.getcwd()

'/home/thomy/Documents'

En el ejemplo pasado, el directorio de trabajo actual estaba ubicado en `'/home/thomy/Google Drive/ITM/Materias/CursoPythonEgresados/JupiterNotebooks/Proyectos/JupyterNotebooks'
`. Finalmente, lo cambiamos al directorio `'/home/thomy/Documents'`. Python mostrará un error si cambiamos el directorio a una ubicación inexistente: 

In [11]:
os.chdir('C:\\EsteDiretorioNoExiste')

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\EsteDiretorioNoExiste'

## Rutas absolutas vs rutas relativas

Existen dos maneras de especificar una ruta de un archivo:

+ La ruta absoluta, la cual siempre empieza con un directorio raíz.
+ La ruta relativa, la cual hace referencia al directorio de trabajo actual.

Adicionalmente, existen los directorios *dot* `.` y *dot-dot* `..`. No son directorios reales, son nombres especiales que pueden ser usados en las rutas. Un único punto como nombre de directorio es un atajo para “este directorio”. Dos puntos seguidos hacen referencia al directorio madre o directorio principal que lo contiene.

La siguiente figura muestra algunos directorios y archivos cuando el directorio actual es fijado en `C:\bacon`. Se puede apreciar la ruta absoluta y la ruta relativa para cada archivo y directorio.


<img src="https://i.imgur.com/qtcdER5.jpg" width="600">

El `.\` al principio de la ruta relativa es opcional. Por ejemplo, `.\spam.txt` y `spam.txt` hacen referencia al mismo archivo.

## Creando directorios nuevos con `os.makedirs()`

Nuestros programas pueden crear nuevos directorios gracias a la función `os.makedirs()`. Veamos cómo funciona.

In [9]:
# Windows
import os
os.makedirs('C:\\delicious\\walnut\\waffles')

In [7]:
# Linux
import os
os.makedirs('/home/thomy/delicious/walnut/waffles')

In [10]:
os.chdir('/home/thomy/delicious/walnut/waffles')
os.getcwd()

'/home/thomy/delicious/walnut/waffles'

Las anteriores instrucciones crearon no sólo el directorio `C:\Delicious` sino también el folder `walnut` dentro de este. De la misma manera creó el directorio `waffles` dentro de `C:\delicious\walnut\`. De esta manera, `os.makedirs()` se asegura de crear todos los directorios intermedios para que la ruta completa exista.

<img src="https://i.imgur.com/YPq7Jxd.jpg" width="200">

# El módulo `os.path()`

El módulo `os.path` contiene muchas funciones útiles relacionadas con los nombres de archivos y sus rutas. Por ejemplo ya hemos usado `os.path.join()` para construir rutas de forma tal que trabajen en diferentes sistemas operativos. Como `os.path` es un módulo dentro del módulo `os`, lo podemos importar simplemente corriendo `import os`. Podemos volver a esta unidad cada vez que nuestros programas precisen trabajar con archivos, nombres de archivos o rutas. La documentación completa para el módulo `os.path` se puede encontrar en http://docs.python.org/3/library/os.path.html.

## Manejando rutas absolutas y rutas relativas

El módulo `os.path` suministra funciones para retornar la ruta absoluta de rutas relativas, y para verificar que una ruta dada es absoluta.

+ Llamar `os.path.abspath(ruta)` retornará una cadena con la ruta absoluta de su argumento. Esta es una manera fácil de convertir una ruta relativa en una absoluta.
+ Llamar `os.path.isabs(ruta)` retornará `True` si el argumento es una ruta absoluta, de lo contrario, retornará `False`.
+ Llamar `os.path.relpath(ruta, inicio)` retornará una cadena correspondiente a la ruta relativa desde el *inicio* hasta *ruta*. Si *inicio* no es suministrado, el directorio de trabajo actual es usado como *inicio*. Veamos algunos ejemplos:


In [16]:
os.path.abspath('.')

'/home/thomy/anaconda3/bin'

In [18]:
os.chdir('/home/thomy/')
os.path.abspath('./projects')

'/home/thomy/projects'

In [19]:
os.path.isabs('.')

False

In [21]:
os.path.isabs(os.path.abspath('.'))

True

**Nota: Dado que sus computadores tienen archivos y carpetas diferentes al mío, en ocasiones no serán capaces de reproducir los resultados de los ejemplos aquí mostrados. No obstante, le sugiero realicen los ejemplos con sus propios directorios y archivos. **

Ingresamos la siguiente llamada a `os.path.relpath()` en el shell interactivo.

In [ ]:
os.path.relpath('C:\\Windows', 'C:\\')

In [24]:
os.path.relpath('/home/thomy/delicious/walnut/waffles', '/home')

'thomy/delicious/walnut/waffles'

In [25]:
os.path.relpath('/home/thomy/', '/home/thomy/delicious/walnut/waffles')

'../../..'

Llamar `os.path.dirname(ruta)` retornará una cadena de cacarteres con todos los directorio que hay antes del último slash en la ruta del argumento. Llamar `os.path.basename(ruta)` retornará una cadena de caracteres todo aquello que está después del último slash de la ruta del argumento.

<img src="https://i.imgur.com/KUDdnjS.png" width="350">

In [ ]:
# Windows
path = 'C:\\Windows\\System32\\calc.exe'

os.path.dirname(path)

In [ ]:
# Windows
os.path.basename(path)

In [30]:
# Linux
path = '/usr/bin/kate'
os.path.dirname(path)

'/usr/bin'

In [31]:
# Linux
os.path.basename(path)

'kate'

Si necesitamos el *dir name* y el *base name* de la ruta establecida como argumento simultáneamente, podemos llamar `os.path.split()` para obtener una tupla con estas dos cadenas.

In [ ]:
# Windows
rutaCalculadora = 'C:\\Windows\\System32\\calc.exe'
os.path.split(rutaCalculadora)

In [33]:
# Linux
rutaKate = '/usr/bin/kate'
os.path.split(rutaKate)

('/usr/bin', 'kate')

Nótese que pudimos haber obtenido la misma tupla si llamamos `os.path.dirname()` y `os.path.basename()` colocando sus valores de retorno en una tupla:

In [ ]:
# Windows
rutaCalculadora = 'C:\\Windows\\System32\\calc.exe'
(os.path.dirname(rutaCalculadora), os.path.dirname(rutaCalculadora))

In [36]:
# Linux
rutaKate = '/usr/bin/kate'
(os.path.dirname(rutaKate), os.path.basename(rutaKate))

('/usr/bin', 'kate')

`os.path.split()` es un buen atajo si necesitamos ambos valores. Adicionalmente, nótese que `os.path.split()` toma una ruta de archivo y NO devuelve una lista de cadenas correspondientes a cada directorio de tal ruta. Par tal propósito, debemos usar el método de las cadenas `split()` sobre `os.sep`. Recordemos que la variable `os.sep` es definida por el separador de directorios de acuerdo al sistema operativo que corre el programa.

In [ ]:
# Windows
rutaCalculadora.split(os.path.sep)

In [37]:
# Linux
rutaKate.split(os.path.sep)

['', 'usr', 'bin', 'kate']

El método para las cadena `split()` funcionará para cualquier sistema operativo si le pasamos como argumento `os.path.sep`.

## Encontrando tamaños de archivo y el contenido de los directorios

Una vez tenemos las herramientas para manipular las rutas de los archivos, podemos empezar a recoger información sobre archivos y directorios específicos. El módulo `os.path` suministra funciones para encontrar el tamaño de un archivo en bytes y los archivos contenidos en un directorio dado.

+ Llamar `os.path.getsize(ruta)` retornará el tamaño en bytes del archivo establecido en la ruta.
+ Llamar `os.listdir(path)` retornará una lista de cadenas de caracteres correspondientes al nombre de cada uno de los archivos de la ruta establecida como argumento . Nótese que este método se encuentra en el módulo `os` y no en `os.path`

Esto es lo que obtenemos cuando probamos estas funciones:


In [ ]:
# Windows
os.path.getsize('C:\\Windows\\System32\\calc.exe')

In [38]:
# Linux
os.path.getsize('/usr/bin/kate')

836984

In [ ]:
# Windows
os.listdir('C:\\Windows\\System32')

In [39]:
# Linux
os.listdir('/usr/bin')

['globaltime',
 'cut',
 'gpg-agent',
 'ppmqvga',
 'parec',
 'makeglossaries',
 'mmcli',
 'llvm-extract-3.8',
 'ksysguard',
 'pwdx',
 'xdg-desktop-menu',
 'fc-validate',
 'stream-im6',
 'ebook-edit',
 'meinproc4_simple',
 'foo2slx',
 'obexctl',
 'ri2.3',
 'genisoimage',
 'apt-add-repository',
 'konqueror',
 'dh_gencontrol',
 'depythontex',
 'wc',
 'aclocal-1.15',
 'cb_console_runner',
 'spd-conf',
 'gnome-keyring-daemon',
 'gsdj500',
 'assistant',
 'xetex',
 'prezip-bin',
 'explode.py',
 'xfce4-popup-notes',
 'nl',
 'dh_installifupdown',
 'compose',
 'pl2link',
 'yuvtoppm',
 'lslogins',
 'ucf',
 'kscreen-doctor',
 'ptargrep',
 'display-im6',
 'opentool',
 'xcutsel',
 'orage',
 'latexdiff-git',
 'evim',
 'systemd-analyze',
 'ebook-convert',
 'gftopk',
 'join',
 'kscreen-console',
 'linkicc',
 'loffice',
 'gifmaker',
 'pgmhist',
 'chacl',
 'pamstack',
 'py3clean',
 'sty2dtx',
 'kwriteconfig',
 'dh_builddeb',
 'pstopnm',
 'texluajitc',
 'rendercheck',
 'ksysguardd',
 'strip',
 'patgen',
 '

Como podemos ver, el editor de texto Kate en mi computador tiene un tamaño de 836984 bytes y tengo muchos archivos dentro de `’/usr/bin’`. Si quiero encontrar el tamaño total de todos los archivos en este directorio, puedo usar `os.path.getsize(ruta)` y `os.listdir(ruta)` juntos de la siguiente manera.

In [41]:
totalSize = 0
for filename in os.listdir('/usr/bin'):
    totalSize = totalSize + os.path.getsize(os.path.join('/usr/bin', filename))
    
print(totalSize)

815953321


En cada iteración sobre el nombre de archivo `filename` en el directorio `/usr/bin` , la variable `totalSize` es incrementada en el tamaño de cada uno de los archivos. Nótese cómo al llamar `os.path.getsize()`, usamos `os.path.join()` para unir la ruta del directorio con el nombre del archivo. El entero que `os.path.getsize()` retorna es sumado a la variable `totalSize`. Al terminar todas las iteraciones, imprimo el valor acumulado de `totalSize`.

## Verificando la validez de la ruta

Muchas funciones de Python producirán un mensaje de error si ingresamos una ruta que no existe. El módulo `os.path` suministra funciones para verificar si la ruta dada existe y si se trata de un archivo o un directorio.

+ Llamar `os.path.exists(ruta)` retornará un `True` si el archivo o el directorio ingresado en la ruta como argumento, existe. Si no existe, retornará un `False`
+ Llamar `os.path.isfile(ruta)` retornará `True` si la ruta del argumento existe y es un archivo, de lo contrario retornará `False`.
+ Llamar `os.path.isdir(ruta)` retornará `True` si la ruta del argumento existe y es un directorio, de lo contrario retornará `False`.

Veamos algunos ejemplos:

__Windows__

In [ ]:
os.path.exists('C:\\Windows')

In [ ]:
os.path.exists('C:\\directorioFalso')

In [ ]:
os.path.isdir('C:\\Windows\\System32')

In [ ]:
os.path.isfile('C:\\Windows\\System32')

In [ ]:
os.path.isdir('C:\\Windpws\\System32\\calc.exe')

In [ ]:
os.path.isfile('C:\\Windpws\\System32\\calc.exe')

__Linux__

In [43]:
os.path.exists('/usr/bin')

True

In [44]:
os.path.exists('/DirectorioInexistente/')

False

In [45]:
os.path.isdir('/usr/bin')

True

In [48]:
os.path.isfile('/usr/bin')

False

In [47]:
os.path.isdir('/usr/bin/kate')

False

In [49]:
os.path.isfile('/usr/bin/kate')

True

De la misma manera, podemos determinar si hay un DVD o una USB conectada al computador mediante la función `os.path.exists()`. Por ejemplo si quiero determinar si hay algo en la unidad *E:\* de mi computador, podría hacerlo de la siguiente manera.

In [ ]:
os.path.exists('D:\\')

# El proceso de lectura/escritura de archivos

Una vez que nos sintamos cómodos trabajando los directorios, rutas relativas y rutas absolutas, podemos especificar la ubicación de los archivos  para leer y escribir sobre ellos. Las funciones que veremos en las siguientes secciones se aplicarán a archivos de texto plano. Los archivos de texto plano contienen únicamente caracteres básicos de texto y no incluyen información sobre el tipo, tamaño de letra o color. 

Los archivos con extensión *.txt* o scripts de Python con extensión *.py* son ejemplos de archivos de texto plano. Estos pueden ser abiertos con el Block de Notas en Windows o cualquier aplicación de edición de texto. Nuestros programas podrán leer fácilmente el contenido de archivos de texto plano y tratar su contenido como una cadena de caracteres ordinaria.

Los archivos binarios son otro tipo de archivos como los documentos de procesamiento de palabras, PDFs, hojas de cálculo y ejecutables. Si abrimos un archivo binario en el Block de Notas, lucirá como un montón de caracteres sin sentido.

<img src="https://i.imgur.com/5UFYdHT.jpg" width="500">

Dado que cada archivo binario se abre de una manera particular, este módulo no entrará a leer o escribir archivos binarios directamente.

Dado que cada tipo de archivo binario se abre de una forma determinada, en este capitulo no profundizaremos en la lectura y escritura de archivos binarios directamente. Afortunadamente, existen varios módulos que nos permiten trabajar con archivos binarios de forma fácil. 

Existen tres pasos para lectura y escritura de archivos en Python:
1. Llamar la función `open()` para retornar un objeto `File`.
2. Llamar los métodos `read()` y `write()`del objeto `File`.
3. Cerrar el archivo llamando el método `close()` del objeto `File`.

## Abriendo archivos con la función `open()`

Para abrir un archivo con la función `open()`, debemos pasarle una cadena como argumento indicando la ruta del archivo que deseamos abrir; puede ser la ruta absoluta o la relativa. La función `open()` siempre retornará un objeto `File`.

Probemos esto creando un archivo de texto llamado `hello.txt` usando el Block de Notas. Escribamos `Hola Mundo!` como contenido de este archivo.  Luego lo guardamos en el directorio *C:\Users\Su_Nombre_de_Usuario\hello.txt*

In [ ]:
# Windows
helloFile = open('C:\\Users\\Su_Nombre_de_Usuario\\hello.txt')

In [2]:
# Linux
helloFile = open('/home/thomy/Documentos/hello.txt')

Asegúrese de reemplazar *Su_Nombre_de_Usuario* por el nombre de usuario de su computador.

Ambos comandos abrirán el archivo en modo “Lectura de texto plano” o simplemente *modo lectura*. Cuando un archivo se abre en modo lectura, Python nos permiten acceder al archivo sólo para leer sus datos, no podemos escribir o modificarlo de ninguna manera. El modo lectura es el modo por defecto para archivos que abrimos en Python. No obstante, si no confiamos en el modo por defecto que tiene Python, podemos especificar explícitamente el modo lectura pasando el valor `'r'` como segundo argumento a la función `open()` de la siguiente manera: `open(‘C:\Users\Su_Nombre_de_Usuario\hello.txt’, 'r')`.

El llamado a `open()` retornará un objetivo `File`. Este objeto representa un archivo en el computador; podemos entender el objeto `File` como otro tipo de dato en Python como las listas o diccionarios que ya conocemos. En el ejemplo anterior almacenamos el objeto `File` en una variable llamada `helloFile`. Ahora, cuando deseemos leer o escribir sobre el archivo, lo podemos hacer usando los métodos disponibles para el objeto en `helloFile`.

## Leyendo el contenido de los archivos

Ahora que tenemos el objeto `File`, podemos empezar a leerlo.  Si deseamos leer el documento completo como una cadena de caracteres, podemos usar el método `read()` del objeto `file`. Continuando con el el objeto `File` *hello.txt* que almacenamos en la variable `helloFile`, ingresemos las siguientes instrucciones en el shell interactivo:

In [3]:
helloContent = helloFile.read()
helloContent

'Hola mundo!\n'

Si pénsamos en el contenido de un archivo como un único valor de cadena de caracteres, el método `read()` retorna la cadena almacenada en dicho archivo. Alternativamente, podemos usar el método `readlines()` para obtener una lista de cadenas del archivo, una cadena por cada línea de texto. Por ejemplo, creemos un archivo texto llamado *sonnet29.txt* en el mismo directorio donde tenemos *hello.txt* y escribamos las siguientes líneas.

When, in disgrace with fortune and men's eyes,  
I all alone beweep my outcast state,  
And trouble deaf heaven with my bootless cries,  
And look upon myself and curse my fate,  

Asegúrense de separar las cuatro líneas con cambios de línea, luego ingresemos lo siguiente en el shell interactivo.

In [7]:
sonnetFile = open('/home/thomy/Documentos/sonnet29.txt')
sonnetFile.readlines()

["When, in disgrace with fortune and men's eyes,\n",
 'I all alone beweep my outcast state,\n',
 'And trouble deaf heaven with my bootless cries,\n',
 'And look upon myself and curse my fate,\n']

Nótese que cada cadena finaliza con el caracter de cambio de línea. Una lista de cadenas es usualmente más fácil de trabajar que una única cadena de catracteres.

## Escribiendo archivos

Python nos permite escribir sobre archivos de manera similar a cómo la función `print()` imprime en pantalla. No podemos escribir sobre un archivo que hayamos abierto en modo lectura. Sin embargo, podemos abrirlo en modo *escritura* o en modo *append*.

El modo escritura sobre-escribirá el archivo existente y empezara desde a cero a modificarlo, de la misma manera que sobre-escribimos el valor de una variable. Pasarle `'w'` como segundo argumento a `open()` abrirá el archivo en modo escritura. Por otro lado, el modo append, insertará texto al final del archivo, podemos pensar en esto como insertar al final de una lista un nuevo valor, en lugar de sobre-escribirla. Pasarle `'a'` como segundo argumento a `open()`, abrirá el archivo en modo append.

Si el nombre de archivo pasado a `open()` no existe, ambos modos crearán un archivo en blanco.

Luego de leer o escribir sobre un archivo debemos llamar el método `close()` antes de abrir el archivo nuevamente.

Veamos estos conceptos integrados en un ejemplo:

In [3]:
baconFile = open('/home/thomy/Documentos/bacon.txt', 'w')
baconFile.write('Hello world!\n')

13

In [2]:
baconFile.close()

In [13]:
baconFile = open('/home/thomy/Documentos/bacon.txt', 'a')
baconFile.write('Bacon is not a vegetable.')

25

In [14]:
baconFile.close()

In [17]:
baconFile = open('/home/thomy/Documentos/bacon.txt')
content = baconFile.read()
baconFile.close()
print(content)

Hello world!
Bacon is not a vegetable.


Primero, abrimos *bacon.txt* en modo escritura. Dado que *bacon.txt* no existe, Python lo crea. Al llamar `write()` sobre el archivo abierto y pasarle la cadena `'Hello world!\n'`, escribimos sobre el archivo y Python retorna el número de caracteres escritos, incluyendo la nueva línea. Luego cerramos el archivo.

Para agregar texto al archivo existente en lugar remplazar el contenido que acabamos de escribir, abrimos el archivo en modo append y escribimos `'Bacon is not a vegetable'` y lo cerramos. Finalmente, para imprimir el contenido del archivo en pantalla, abrimos el archivo en su modo por defecto, es decir, en modo lectura, llamamos a la función `read()`, almacenamos el objeto `File` resultante en la variable `content`, cerramos el archivo e imprimimos la variable.

Nótese que el método `write()` no agrega automáticamente el carácter de cambio de línea al final de la cadena como lo hace la función `print()`. Debemos agrehar este carácter manualmente.

# Guardando variables con el método `shelve`

Podemos guardar variables en nuestros programas de Python en archivos binarios *shelf* usando el módulo `shelve`. De esta manera, nuestros programas podrán restaurar datos para nuestras variables desde el disco duro. El módulo `shelve` nos permite agregar características de guardado y apertura a nuestros programas. Por ejemplo, si abrimos un programas y definimos algunos ajustes de configuración, podemos guardar esos ajustes en un archivo shelf para luego hacer que el programa los cargue la próxima vez que lo corramos. Ingresemos lo siguiente en el shell interactivo.

In [5]:
import shelve, os

os.getcwd()
os.chdir('/home/thomy/Documentos/')

shelfFile = shelve.open('myData')
cats = ['Zophie', 'Pooka', 'Simon']
shelfFile['cats'] = cats
shelfFile.close()

In [6]:
import os
os.getcwd()
os.listdir()

['myData.dir',
 'bacon.txt',
 'myData.bak',
 'sonnet29.txt',
 'hello.txt',
 'myData.dat',
 'Project.docx']

Para leer y escribir datos usando el módulo `shelve`, debemos importarlo primero. Llamamos `shelve.open()` y le pasamos un nombre de archivo para luego almacenar el valor shelf retornado en una variable. Podemos realizar cambios al valor shelf como si fuese un diccionario. Cuando terminemos, llamamos el método `close()` sobre el valor shelf. En el ejemplo anterior, nuestro valor shelf es almacenado en `shelfFile`. Creamos la lista `cats` y escribimos `shelfFile['cats'] = cats` para almacenar la lista en `shelfFile` como valor asociado a la llave `'cats'` (tal y como funciona un diccionario). Luego llamamos `close()` sobre `shelfFile`.

Luego de correr el código anterior, veremos tres nuevos archivos en el directorio de trabajo actual: *mydata.bak*, *mydata.dat* y *mydata.dir*. Estos archivos binarios contienen la información almacenada en el *shelf*. El formato de estos archivos no es importante; sólo necesitamos saber qué hace el módulo `shelve` y no cómo lo hace. El módulo nos libera de preocuparnos sobre el cómo se almacena los datos de nuestro programa en un archivo.

Nuestros programas pueden usar el módulo `shelve` para, posteriormente, abrir y recuperar los datos de  los archivos shelf. Los valores shelf no necesitan que se abran en modo escritura o lectura, pueden trabajar con ambos modos una vez abiertos. Veamos el siguiente ejemplo:

In [7]:
shelfFile = shelve.open('myData')
type(shelfFile)

shelve.DbfilenameShelf

In [8]:
shelfFile['cats']

['Zophie', 'Pooka', 'Simon']

In [9]:
shelfFile.close()

En el ejemplo anterior abrimos el archivo `shelf` para verificar que nuestros datos quedaron almacenados correctamente. Al ingresar `shelfFile['cats']`, Python retorna la misma lista que almacenamos previamente. De esta manera nos aseguramos que la lista fue almacenada existosamente. Por último cerramos el archivo llamando `close()`.

De la misma manera que los diccionarios, el tipo de dato shelf tiene métodos `keys()` y `values()` que retornarán valores similares a las listas de las llaves y los valores en el shelf. Dado que estos métodos retornan valores similares a las listas, debemos pasarlos a la función `list()` para obtener su valor de tipo lista. Veamos el siguiente ejemplo:

In [11]:
shelfFile = shelve.open('myData')
list(shelfFile.keys())

['cats']

In [12]:
list(shelfFile.values())

[['Zophie', 'Pooka', 'Simon']]

In [13]:
shelfFile.close()

El texto plano es útil para crear archivos que leeremos en un editor de texto como Block de Notas, pero si deseamos almacenar datos desde nuestros programas de Python, resulta más conveniente usar el módulo `shelve`.

# Guardando variables con la función `pprint.pformat()`

Recordemos que la función `pprint()` imprime estéticamente  el contenido de un diccionario en pantalla, mientras que la función `pprint.pformat()` retorna el mismo texto como cadenas de caracteres en lugar de imprimirlo. No sólo esta cadena es formateada para su fácil lectura, sino que es una forma de código sintácticamente correcta de Python. Digamos que tenemos un diccionario almacenado en una variable, y deseamos guardar esta variable y su contenido para un uso futuro. Usar  `pprint.pformat()` nos dará una cadena que podemos escribir en un archivo *.py*. Este archivo será nuestro propio módulo que importaremos cuando necesitemos la variable allí almacenada. Veamos un ejemplo:

In [15]:
import pprint
cats = [{'name': 'Zophie', 'desc': 'chubby'}, {'name': 'Pooka', 'desc': 'fluffy'}]
pprint.pformat(cats)

"[{'desc': 'chubby', 'name': 'Zophie'}, {'desc': 'fluffy', 'name': 'Pooka'}]"

In [20]:
fileObj = open('myCats.py', 'w')
fileObj.write('cats = ' + pprint.pformat(cats) + '\n')

83

In [21]:
fileObj.close()

En el ejemplo anterior importamos `pprint` ya que nos permite usar `pprint.pformat()`. Tenemos una lista de diccionarios almacenada en en la variable `cats`. Para mantener la lista en la variable `cats` incluso después de cerrar el jupyter notebook o el IDE, usamos `pprint.pformat()` para retornarla como cadena de caracteres. Una vez tenemos los datos en `cats` como cadena, resulta fácil escribirla en una archivo llamado *myCats.py.*

Los módulos que importa el estamento `import`, de hecho son scripts en Python, cuando la cadena retornada por `pprint.pformat()` es guardada en un archivo *.py*, éste es un módulo que puede ser importado justo como también otro.

Finalmente, dado que los scripts en Python son sólo texto con la extensión *.py*, nuestros progreamas en pueden generar incluso otros programas. Luego podemos importar dicho archivos a manera de scripts:


In [22]:
import myCats
myCats.cats

[{'desc': 'chubby', 'name': 'Zophie'}, {'desc': 'fluffy', 'name': 'Pooka'}]

In [23]:
myCats.cats[0]

{'desc': 'chubby', 'name': 'Zophie'}

In [25]:
myCats.cats[0]['name']

'Zophie'

El beneficio de crear un archivo *.py*, a diferencia de guardar variables usando el módulo `shelve`, es que debido a que el primero es un archivo de texto, el contenido de este puede ser leído o modificado por cualquiera con un editor de texto. Para la mayoría de aplicaciones, sin embargo, guardar los datos usando el módulo `shelve` es la forma preferida para guardar variables en un archivo. Sólo tipo de datos básicos como enteros, números de punto flotante, cadenas, listas y diccionarios, pueden ser almacenados como texto simple. Los objetos `File`, por ejemplo, no pueden ser codificados como texto.

# Proyecto: Generando archivos aleatorios para evaluaciones

Supongamos que usted es un profesor de geografía con 35 estudiantes en su clase y desea realizar in quiz sobre las capitales de los departamentos de Colombia. Nuestros estudiantes tienden a hacer trampa y no podemos confiar en ellos. Nos gustaría organizar aleatoriamente el orden de las preguntas de forma tal que cada quiz sea único, haciéndolo imposible para que los estudiantes puedan copiar sus respuestas de otro. Hacer esto a mano puede ser un poco tedioso y demandante en tiempo. Afortunadamente, sabemos un poco de Python.
He aquí lo deseamos que nuestro programa realice:

+ Crear 35 quizzes diferente.
+ Crear 32 preguntas de selección multiple para cada quiz en orden aleatorio.
+ Suministrar la respuesta correcta y tres respuestas incorrectas de forma aleatoria.
+ Escribir los quizzes en 35 archivos.
+ Escribir las respuestas de los 35 archivos de texto.

Esto significa que nuestro programa hará lo siguiente:

+ Almacenar los departamentos y las capitales en un diccionario
+ Llamar a las funciones `open()`, `write()` y `close()` para el quiz y el archivo de respuestas.
+ Usar `random.shuffle()` para ordenar aleatoriamente las preguntas y las opciones de selección multiple

## Paso1: Almacenar los datos del quiz en un diccionario

El primer paso es crear el esqueleto y llenarlo con los datos del examen. Creemos un archivo llamado *randomQuizGenerator.py* y hagamos que luzca de la siguiente manera.

In [1]:
import sys

print(sys.executable)

/home/thomy/anaconda3/bin/python


In [ ]:
#! /home/thomy/anaconda3/bin/python
# randomQuizGenerator.py - Crea los quizzes con preguntas y respuestas alatorias
# así como el archivo de respuestaser key.

import random

# Información del quiz. 
# Diccionaro cuyas llaves son los departamentos y los valores las capitales

capitales = {'Amazonas': 'Leticia',
'Antioquia': 'Medellín',
'Arauca': 'Arauca',
'Atlántico': 'Barranquilla',
'Bolívar': 'Cartagena',
'Boyacá': 'Tunja',
'Caldas': 'Manizales',
'Caquetá': 'Florencia',
'Casanare': 'Yopal',
'Cauca': 'Popayán',
'Cesar': 'Valledupar',
'Chocó': 'Quibdó',
'Córdoba': 'Montería',
'Cundinamarca': 'Bogotá',
'Guainía': 'Puerto Inírida',
'Guaviare': 'San José del Guaviare',
'Huila': 'Neiva',
'La Guajira': 'Riohacha',
'Magdalena': 'Santa Marta',
'Meta': 'Villavicencio',
'Nariño': 'Pasto',
'Norte de Santander': 'Cúcuta',
'Putumayo': 'Mocoa',
'Quindio': 'Armenia',
'Risaralda': 'Pereira',
'San Andres y Providencia': 'San Andres',
'Santander': 'Bucaramanga',
'Sucre': 'Sincelejo',
'Tolima': 'Ibagué',
'Valle del Cauca': 'Cali',
'Vaupés': 'Mitú',
'Vichada': 'Puerto Carreño'}

# Generar los 35 archivos del quiz

for quizNum in range(35):
    # TODO: Crear el quiz y los archivos de respuestas
    
    #TODO: Escribir el encabezado para el quiz
    
    #TODO: Organizar aleatoriamente el orden de los departamentos
    
    #TODO Iterar a través de los 32 departamentos creando una pregunta para cada uno

Dado que nuestro programa ordenará aleatoriamente las preguntas y respuestas, necesitaremos importar el módulo `random` para usar sus funciones. La variable `capitales` contiene un diccionario con todos los departamentos de Colombia como llaves y sus respectivas capitales como valores. Y dado que, deseamos crear 35 quizzes, el código que de hecho genera el qiuz y el archivo de respuestas (marcado por el momento como comentarios `TODO`), iría dentro del ciclo `for` que iterará 35 veces (Este número puede cambiar dependiendo de la cantidad de quizzes a generar).

## Paso 2: Crear el archivo del quiz y ordenar aleatoriamente las preguntas

Ahora es momento de empezar a llenar los `TODO`.
El código en el ciclo iterará 35 veces, una vez por cada quiz, de forma tal que nos preocupemos sólo por un quiz al tiempo dentro del ciclo de repetición. Primero debemos crear el archivo quiz. Éste debe tener un nombre de archivo único y también debe tener algún tipo de encabezado estándar con espacios en blanco para que el estudiante llene con su nombre, fecha y grupo. Luego necesitamos obtener una lista de departamentos en orden aleatorio, el cual será utilizado posteriormente en la creación de las preguntas y las respuestas para el quiz.

Ingresemos las siguientes lpineas de código al archivo *randomQuizGenerator.py*.

In [ ]:
#! /home/thomy/anaconda3/bin/python
# randomQuizGenerator.py - Crea los quizzes con preguntas y respuestas alatorias
# así como el archivo de respuestaser key.

import random

# Información del quiz. 
# Diccionaro cuyas llaves son los departamentos y los valores las capitales

capitales = {'Amazonas': 'Leticia',
'Antioquia': 'Medellín',
'Arauca': 'Arauca',
'Atlántico': 'Barranquilla',
'Bolívar': 'Cartagena',
'Boyacá': 'Tunja',
'Caldas': 'Manizales',
'Caquetá': 'Florencia',
'Casanare': 'Yopal',
'Cauca': 'Popayán',
'Cesar': 'Valledupar',
'Chocó': 'Quibdó',
'Córdoba': 'Montería',
'Cundinamarca': 'Bogotá',
'Guainía': 'Puerto Inírida',
'Guaviare': 'San José del Guaviare',
'Huila': 'Neiva',
'La Guajira': 'Riohacha',
'Magdalena': 'Santa Marta',
'Meta': 'Villavicencio',
'Nariño': 'Pasto',
'Norte de Santander': 'Cúcuta',
'Putumayo': 'Mocoa',
'Quindio': 'Armenia',
'Risaralda': 'Pereira',
'San Andres y Providencia': 'San Andres',
'Santander': 'Bucaramanga',
'Sucre': 'Sincelejo',
'Tolima': 'Ibagué',
'Valle del Cauca': 'Cali',
'Vaupés': 'Mitú',
'Vichada': 'Puerto Carreño'}

# Generar los 35 archivos del quiz
for quizNum in range(35):
    # Creamos el archivo del quiz y el de respuestas
    quizFile = open('capitalsquiz%s.txt' % (quizNum + 1), 'w')
    answerKeyFile = open('capitalsquiz_answers%s.txt' % (quizNum + 1), 'w')
    
    # Escribimos el encabezado del quiz
    quizFile.write('Nombre:\n\nFecha:\n\nGrupo:\n\n')
    quizFile.write((' ' * 20) + 'Quiz: Capitales de los departamentos de Colombia (Formulario %s)' % (quizNum + 1))
    quiz.File.write('\n\n')
    
    # Organizamos aleatoriamente los departamentos
    departamentos = list(capitales.keys())
    random.shuffle(departamentos)
    
    # TODO: Iterar sobre los 32 departamentos creando una pregunta para cada uno.

El nombre de archivo para los quizzes será *capitalsquiz[N].txt*, donde [N] es un único número para el quiz que viene dado por la variable `quizNum`, el contador del ciclo `for`. El archivo de respuestas para *capitalsquiz[N].txt*, será almacenado en un archivo llamado *capitalsquiz_answers[N].txt*. Cada vez que se itera en el ciclo, el marcador de posición `%s` en `'capitalsquiz%s.txt'` y `'capitalsquiz_answers%s.txt'`, será reemplazado por `(quizNum + 1)`, así el primer quiz y el archivo de respuestas se llamarán *capitalsquiz1.txt* y *capitalsquiz_answers1.txt*, respectivamente. Estos archivos serán creados mediante el llamado a la función `open()`, con segundo argumento `'w'` para abrirlos en modo escritura.

El estamento `write()`, crea el encabezado del quiz para que los estudiantes diligencien con su nombre, fecha y grupo. Finalmente, una lista aleatoria de los departamentos de Colombia es creada con la ayuda de la función `random.shuffle()`, la cual organiza aleatoriamente los valores en una lista que le es pasada como argumento.

## Paso 3: Crear las opciones de respuesta

Ahora, necesitamos generar las opciones de respuestas para cada pregunta de selección múltiple de A hasta D. Para esto, necesitamos crear otro ciclo de cara a generar el contenido para cada una de las 32 preguntas del quiz. Luego, habrá un tercer ciclo anidado para generar las opciones de selección múltiple para cada pregunta. 

In [ ]:
#! /home/thomy/anaconda3/bin/python
# randomQuizGenerator.py - Crea los quizzes con preguntas y respuestas alatorias
# así como el archivo de respuestaser key.

import random

# Información del quiz. 
# Diccionaro cuyas llaves son los departamentos y los valores las capitales

capitales = {'Amazonas': 'Leticia',
'Antioquia': 'Medellín',
'Arauca': 'Arauca',
'Atlántico': 'Barranquilla',
'Bolívar': 'Cartagena',
'Boyacá': 'Tunja',
'Caldas': 'Manizales',
'Caquetá': 'Florencia',
'Casanare': 'Yopal',
'Cauca': 'Popayán',
'Cesar': 'Valledupar',
'Chocó': 'Quibdó',
'Córdoba': 'Montería',
'Cundinamarca': 'Bogotá',
'Guainía': 'Puerto Inírida',
'Guaviare': 'San José del Guaviare',
'Huila': 'Neiva',
'La Guajira': 'Riohacha',
'Magdalena': 'Santa Marta',
'Meta': 'Villavicencio',
'Nariño': 'Pasto',
'Norte de Santander': 'Cúcuta',
'Putumayo': 'Mocoa',
'Quindio': 'Armenia',
'Risaralda': 'Pereira',
'San Andres y Providencia': 'San Andres',
'Santander': 'Bucaramanga',
'Sucre': 'Sincelejo',
'Tolima': 'Ibagué',
'Valle del Cauca': 'Cali',
'Vaupés': 'Mitú',
'Vichada': 'Puerto Carreño'}

# Generar los 35 archivos del quiz
for quizNum in range(35):
    # Creamos el archivo del quiz y el de respuestas
    quizFile = open('capitalsquiz%s.txt' % (quizNum + 1), 'w')
    answerKeyFile = open('capitalsquiz_answers%s.txt' % (quizNum + 1), 'w')
    
    # Escribimos el encabezado del quiz
    quizFile.write('Nombre:\n\nFecha:\n\nGrupo:\n\n')
    quizFile.write((' ' * 20) + 'Quiz: Capitales de los departamentos de Colombia (Formulario %s)' % (quizNum + 1))
    quiz.File.write('\n\n')
    
    # Organizamos aleatoriamente los departamentos
    departamentos = list(capitales.keys())
    random.shuffle(departamentos)
    
# Iteramos a través de los 32 departamentos, creando un pregunta para cada uno
    for questionNum in range(32):
        # Obtenemos las respuestas correctas y las incorrectas
        correctAnswer = capitales[departamentos[questionNum]]
        wrongAnswers = list(capitales.values())
        del wrongAnswers[wrongAnswers.index(correctAnswer)]
        wrongAnswers = random.sample(wrongAnswers, 3)
        answerOptions = wrongAnswers + [correctAnswer]
        random.shuffle(answerOptions)
        
        # TODO: Escribir las preguntas y las opciones de respuestas en el archivo del quiz
        
        # TODO: Escribir las respuestas en un archivo.

La lista de las posibles respuestas incorrectas es un poco más complicado. Se obtienen tras duplicar todos los valores en el diccionario `capitales`, eliminando la respuesta correcta y seleccionando tres valores aleatorios de la lista. La función `random.sample()` nos facilita la realización de esta selección. El primer argumento es la lista desde la cual vamos a empezar a seleccionar; El segundo argumento es el número de valores que deseamos seleccionar. La lista completa de respuestas es la combinación de estas tres respuestas incorrectas con las respuestas correctas. Finalmente, las respuestas necesitan ser organizadas aleatoriamente con el objetivo de que las respuesta correcta no sea siempre la opción D.

## Paso 4: Escribir el contenido de los archivos del quiz y el archivo de respuestas

Todo lo que queda haciendo falta es escribir la pregunta al archivo del quiz y la respuesta al archivo de respuestas. Veamos: 

In [9]:
#! /home/thomy/anaconda3/bin/python
# randomQuizGenerator.py - Crea los quizzes con preguntas y respuestas alatorias
# así como el archivo de respuestaser key.

import random

# Información del quiz. 
# Diccionaro cuyas llaves son los departamentos y los valores las capitales

capitales = {'Amazonas': 'Leticia',
'Antioquia': 'Medellín',
'Arauca': 'Arauca',
'Atlántico': 'Barranquilla',
'Bolívar': 'Cartagena',
'Boyacá': 'Tunja',
'Caldas': 'Manizales',
'Caquetá': 'Florencia',
'Casanare': 'Yopal',
'Cauca': 'Popayán',
'Cesar': 'Valledupar',
'Chocó': 'Quibdó',
'Córdoba': 'Montería',
'Cundinamarca': 'Bogotá',
'Guainía': 'Puerto Inírida',
'Guaviare': 'San José del Guaviare',
'Huila': 'Neiva',
'La Guajira': 'Riohacha',
'Magdalena': 'Santa Marta',
'Meta': 'Villavicencio',
'Nariño': 'Pasto',
'Norte de Santander': 'Cúcuta',
'Putumayo': 'Mocoa',
'Quindio': 'Armenia',
'Risaralda': 'Pereira',
'San Andres y Providencia': 'San Andres',
'Santander': 'Bucaramanga',
'Sucre': 'Sincelejo',
'Tolima': 'Ibagué',
'Valle del Cauca': 'Cali',
'Vaupés': 'Mitú',
'Vichada': 'Puerto Carreño'}

# Generar los 35 archivos del quiz
for quizNum in range(35):
    # Creamos el archivo del quiz y el de respuestas
    quizFile = open('capitalsquiz%s.txt' % (quizNum + 1), 'w')
    answerKeyFile = open('capitalsquiz_answers%s.txt' % (quizNum + 1), 'w')
    
    # Escribimos el encabezado del quiz
    quizFile.write('Nombre:\n\nFecha:\n\nGrupo:\n\n')
    quizFile.write((' ' * 20) + 'Quiz: Capitales de los departamentos de Colombia (Formulario %s)' % (quizNum + 1))
    quizFile.write('\n\n')
    
    # Organizamos aleatoriamente los departamentos
    departamentos = list(capitales.keys())
    random.shuffle(departamentos)
    
# Iteramos a través de los 32 departamentos, creando un pregunta para cada uno
    for questionNum in range(32):
        # Obtenemos las respuestas correctas y las incorrectas
        correctAnswer = capitales[departamentos[questionNum]]
        wrongAnswers = list(capitales.values())
        del wrongAnswers[wrongAnswers.index(correctAnswer)]
        wrongAnswers = random.sample(wrongAnswers, 3)
        answerOptions = wrongAnswers + [correctAnswer]
        random.shuffle(answerOptions)
        
        # Escribamos la pregunta y las opciones de respuesta en el archivo del quiz
        quizFile.write('%s. Cual es la capital de %s?\n' % (questionNum + 1, departamentos[questionNum]))
        
        for i in range(4):
            quizFile.write('    %s. %s\n' % ('ABCD'[i], answerOptions[i]))

        quizFile.write('\n')
            
        # Escribamos las respuestas en un archivo.
        answerKeyFile.write('%s. %s\n' % (questionNum + 1, 'ABCD'[answerOptions.index(correctAnswer)]))
    quizFile.close()
    answerKeyFile.close()

Un ciclo `for` que itera a través de los enteros ubicados en el rango de 0 a 3, escribirá las opciones de respuesta en la lista `answerOptions`. La expresión `'ABCD'[i]` trata la cadena como un arreglo y se evaluara a  `'A'`, `'B'`, `'C'`, y `'D'` para iteración.

En la línea final, la expresión `answerOptions.index(correctAnswer)` encontrará el índice entero de la respuesta correcta en las opciones ordenadas aleatoriamente, y `'ABCD'[answerOptions.index(correctAnswer)]` se evaluará a la letra de la respuesta correcta a ser inscrita en el archivo de respuestas.

Luego de correr el programa, así es como el archivo *capitalsquiz1.txt* lucirá; obviamente sus preguntas y respuestas lucirán diferentes a las mostradas aquí, esto es debido al resultado obtenido por el llamado a la función `random.shuffle()` y su naturaleza aleatoria.

# Proyecto: Multiclipboard

Supongamos que tenemos la tediosa tarea de llenar formularios en una página web o algún software con varios campos de texto. El portapapeles nos salva de estar escribiendo el mismo texto una y otra vez. Desafortunadamente para nuestro propósito, sólo se puede almacenar un solo dato en el portapapeles a la vez. Si tenemos varios fragmentos de texto que necesitamos copiar y pegar, debemos resaltar y copiar una y otra vez para cada fragmento.

Afortunadamente, podemos escribir un programa en Python para llevar un registro de diferentes fragmentos de texto a la vez. A este multiportapapales o *multiclipboard* lo llamaremos *mcb.pyw* (*mcb* es la abreviación de multiclipboard). La extensión *.pyw* significa que Python no mostrará una ventana de terminal cuando corramos el programa.

El programa guardará cada fragmento de texto en el portapapeles bajo un palabra clave. Por ejemplo, cuando corramos `py mcb.pyw save spam`, el contenido actual del portapapeles será almacenado en la palabra clave *spam*. Este texto puede ser cargado posteriormente al portapapeles corriendo `py mcb.pyw spam`. Adicionalmente, si el usuario olvida las palabras clave, pueden correr ` py mcb.pyw list` para copiar todas las palabras clave al portapapeles.

He aquí lo que hace el programa:

+ El argumento de la línea de comando es verificado
+ Si el argumento es `save`, el contenido del portapapeles es almacenado en la palabra clave
+ Si el argumento es `list`, todas las palabras clave son copiadas al portapapeles
+ De lo contrario, el texto de la palabra clave es copiado al portapapeles.

Esto significa que:

+ Leemos los argumentos de la línea de comandos desde `sys.argv`
+ Leemos y escribimos en el portapapeles
+ Guardamos y cargamos en un archivo shelf

Si estamos usando Windows, podemos fácilmente, correr este script desde la ventana de ejecutar mediante la creación de un archivo batch llamado *mcb.bat* con el siguiente contenido:

In [ ]:
@pyw.exe C:\thomy\pythonscrpts\mcb.pyw %*

## Paso 1: Comentarios y configuración del shelf

Empecemos por crear un script con el esqueleto definido por algunos comentarios y la configuración básica. Escribamos lo siguiente:

In [ ]:
#! /home/thomy/anaconda3/bin/python
# mcb.pyw - Guarda y carga fragmentos de texto en el portapapeles
# Modo de uso: py.exe mcb.pyw save <keyword> - guarda el portapapeles en la palabra clave.
#              py.exe mcb.pyw <keyword> - carga la palabra clave al portapapeles.
#              py.exe mcb.pyw list - carga todas las palabras clave al portapapeles


import shelve, pyperclip, sys

mcbShelf = shelve.open('mcb')

# TODO: Guarda el contenido del portapapeles

# TODO: Lista las palabras clave y carga su contenido

mcbShelf.close()

Es una práctica común el poner la información acerca del uso de los programas en los comentarios al inicio del código. Si en algún momento olvidamos cómo correr el script, siempre podemos volver a los comentarios para recordarlo. 

En la segunda parte del script importamos los módulos necesarios para nuestro programa. Copiar y pegar en el portapapeles requiere el módulo `pyperclip` y leer los argumentos de la línea de comandos requiere el módulo `sys`. Por otro lado, el módulo `shelve` nos será de utilidad cuando un usuario desee guardar un fragmento de texto en el portapapeles y lo almacenemos como un archivo shelf. Luego cuando el usuario desee copiar de vuelta el texto previamente almacenado en el portapapeles, abriremos el archivo shelf y lo cargaremos en el programa. El archivo shelf será nombrado con el prefijo *mcb*.

## Paso 2: Guardar el contenido del portapapeles con una palabra clave

El programa realiza diferentes tareas dependiendo de si el usuario desea guardar texto en el portapapeles, cargar fragmentos de texto o listar las palabras claves. Empecemos por el principio, escribamos el siguiente código fuente.

In [ ]:
#! /home/thomy/anaconda3/bin/python
# mcb.pyw - Guarda y carga fragmentos de texto en el portapapeles
# Modo de uso: py.exe mcb.pyw save <keyword> - guarda el portapapeles en la palabra clave.
#              py.exe mcb.pyw <keyword> - carga la palabra clave al portapapeles.
#              py.exe mcb.pyw list - carga todas las palabras clave al portapapeles


import shelve, pyperclip, sys

mcbShelf = shelve.open('mcb')

# Guardamos el contenido del portapapeles
if len(sys.argv) == 3 and sys.argv[1].lower() == 'save':
    mcbShelf[sys.argv[2]] = pyperclip.paste()
elif len(sys.argv) == 2:
    # TODO: Listar las palabras clave y cargar el contenido


mcbShelf.close()


Si el primer argumento de la línea de comandos (el cual siempre va a tener índice 1 de la lista `sys.argv`) es `'save'`, el segundo argumento de la línea de comandos debe ser la palabra clave para el contenido actual del portapapeles. La palabra clave será usada como llave para `mcbShelf` y el valor será el texto actual del portapapeles.

Si sólo hay un argumento en la línea de comandos, podemos suponer que es `'list'` o la palabra clave para cargar su contenido en el portapapeles. Podemos implementar el código para esto más tarde, por lo pronto, pongamos un comentario `TODO`.

## Paso 3: Listar las palabras clave y cargar el contenido de la palabra clave

Finalmente, implementemos los casos restantes: El usuario desea cargar el texto del portapapeles a partir de la palabra clave, o desea listar todas las palabras clave disponibles. Hagamos que nuestro script luzca de la siguiente manera:

In [ ]:
#! /home/thomy/anaconda3/bin/python
# mcb.pyw - Guarda y carga fragmentos de texto en el portapapeles
# Modo de uso: py.exe mcb.pyw save <keyword> - guarda el portapapeles en la palabra clave.
#              py.exe mcb.pyw <keyword> - carga la palabra clave al portapapeles.
#              py.exe mcb.pyw list - carga todas las palabras clave al portapapeles


import shelve, pyperclip, sys

mcbShelf = shelve.open('mcb')

# Guardamos el contenido del portapapeles
if len(sys.argv) == 3 and sys.argv[1].lower() == 'save':
    mcbShelf[sys.argv[2]] = pyperclip.paste()
elif len(sys.argv) == 2:
    # Listamos las palabras clave y cargamos el contenido
    if sys.argv[1].lower() == 'list':
        pyperclip.copy(str(list(mcbShelf.keys())))
    elif sys.argv[1] in mcbShelf:
        pyperclip.copy(mcbShelf[sys.argv[1]])

mcbShelf.close()

Si únicamente existe un argumento en la línea de comandos, verificamos primero si es `'list'`. En caso afirmativo, una representación en cadena de caracteres de la lista de las llaves será copiada al portapapeles. El usuario podrá pegar esta lista en un editoir de texto para leerla.

De lo contrario, podemos asumir que el argumento de la lista de comandos es una palabra clave. Si esta existe en `mcbShelf` como llave, cargamos su valor correspondiente al portapapeles.